core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java

```diff
@@ -239,4 +261,62 @@ private IndexMetaData addDefaultUnitsIfNeeded(IndexMetaData indexMetaData) {
         // No changes:
         return indexMetaData;
     }
+
+
+    /**
+     * Checks the mappings of the mappings for compatibility with the current version
+     */
+    private IndexMetaData upgradeMappings(IndexMetaData indexMetaData) {
+        Index index = new Index(indexMetaData.getIndex());
+        Settings settings = indexMetaData.settings();
+        try {
+            SimilarityLookupService similarityLookupService = new SimilarityLookupService(index, settings);
+            // We cannot instantiate real analysis server at this point because the node might not have
+            // been started yet. However, we don't really need real analyzers at this stage - so we can fake it
+            try (AnalysisService analysisService = new FakeAnalysisService(index, settings)) {
+                try (MapperService mapperService = new MapperService(index, settings, analysisService, similarityLookupService, scriptService)) {
+                    for (ObjectCursor<MappingMetaData> cursor : indexMetaData.getMappings().values()) {
+                        MappingMetaData mappingMetaData = cursor.value;
+                        mapperService.merge(mappingMetaData.type(), mappingMetaData.source(), true, true);
+                    }
+                }
+            }
+        } catch (Exception ex) {
+            // Wrap the inner exception so we have the index name in the exception message
+            throw new IllegalArgumentException("unable to upgrade the index [" + indexMetaData.getIndex() + "], reason: [" + ex.getMessage() + "]", ex);
+        }
+        return indexMetaData;
+    }
+
+    /**
+     * Marks index as upgraded so we don't have to test it again
+     */
+    private IndexMetaData markAsUpgraded(IndexMetaData indexMetaData) {
+        Settings settings = Settings.builder().put(indexMetaData.settings()).put(IndexMetaData.SETTING_VERSION_UPGRADED, Version.CURRENT).build();
+        return IndexMetaData.builder(indexMetaData).settings(settings).build();
+    }
+
+    /**
+     * A fake analysis server that returns the same keyword analyzer for all requests
+     */
+    private static class FakeAnalysisService extends AnalysisService {
+
+        private static Analyzer keywordAnalyzer = new KeywordAnalyzer();
```

Comment:
The reason why I added this comment was because I saw it was closed in FakeAnalysisService.close(). So instead should we create an anonymous fake analyzer (that throws an exception in createComponents) and never close it?